# OrthoreoVariant

Code related to viral single nucleotide variant detection in Illumina datasets.

## Description

Code utilized in Ogden lab to analyze viral SNVs found in Illumina datasets.

## Getting Started

### Dependencies

* Analysis was performed on an Ubuntu 20 OS. Please note your available computational resourcing and adjust accordingly.
* Pre-processing: Trimmomatic-0.39
* Alignment: bowtie2, samtools, bbmap
* Variant calling: LoFreq
* Python 3.11+
* Python packages: pandas, os, fnmatch, argparse, pathlib

### Installing

* Code can be downloaded directly from this repo. Bash commands can be adjusted as necessary

### Executing program

* How to run the program
* Step-by-step bullets
```
code blocks for commands
```

## Help

Any advise for common problems or issues.
```
command to run if program contains helper info
```

## Authors

Contributors names and contact info

ex. Dominique Pizzie  
ex. [@DomPizzie](https://twitter.com/dompizzie)

## Version History

* 0.2
    * Various bug fixes and optimizations
    * See [commit change]() or See [release history]()
* 0.1
    * Initial Release

## License

This project is licensed under the [NAME HERE] License - see the LICENSE.md file for details

## Acknowledgments

Inspiration, code snippets, etc.
* [awesome-readme](https://github.com/matiassingers/awesome-readme)
* [PurpleBooth](https://gist.github.com/PurpleBooth/109311bb0361f32d87a2)
* [dbader](https://github.com/dbader/readme-template)
* [zenorocha](https://gist.github.com/zenorocha/4526327)
* [fvcproductions](https://gist.github.com/fvcproductions/1bfc2d4aecb01a834b46)
